### ISSUE TITLE
Dart binding for web use - JS Interop

### ISSUE THREAD

**Issue by pratikpparikh** [rustwasm/wasm-bindgen#2561]
Converting your existing rust code to support the web via web assembly is an efficient and effective way to reuse your code.   Flutter provides a way to code across many client environments.  It would be very helpful if wasm-bindgen supported Flutter using JS interop. 

Provide a Dart option along with side JS and TS will help a lot of Flutter developers to use Rust as the highly reusable programming language that targets all environments in line with flutter.

The current alternative is to use the web option on the generator and handwrite wrapper in HTML code.  Think if you have 20 wasm libraries that you are using the management of those 20 code libs will be a messy task.

**Comment by jonboj** [comment]
Not part of the wasm-bindgen developers, but have made a project with integration toward wasm-bindgen an other with integration toward Dart-JS Interop.

For the Dart/Flutter many platforms and frameworks involved, so combine to many scenarios. Maybe more clarification of which scenario addressed could help. Some parts seem to address a more highlevel/longterm roadmap of wasm and Dart/Flutter.

Below are some opinionated observations.

Normally integration towards wasm is done for performance or/and getting a “better” dev environment/prog language than JS. In the case of performance native ffi with dart:ffi seems faster with Rust on the metal in e.g. Flutter Desktop. For better prog language than JS, why not stay with Dart?

An other observation, Dart/Flutter for the web is based on transpiling with dart2js. The Google Dart team avoids forward looking statements. But with Dart being a GCed language, not wrong to guess that if/when a GC feature https://github.com/WebAssembly/gc is available, a dart2wasm compiler could be developed. With dart2wasm wasm could replace JS as target for Dart on the web. Hence in this case making a Dart integration with wasm-bindgen superfluous.

--

Don’t get this scenario in detail, but looks interesting.
>The current alternative is to use the web option on the generator and handwrite wrapper in HTML code. Think if you have 20 wasm libraries that you are using the management of those 20 code libs will be a messy task.

From Dart invoking into wasm modules? Something like Dart -> JS Interop -> wasm-bindgen -> wasm
Following https://github.com/rustwasm/wasm-bindgen/tree/master/examples/hello_world
This will need a Dart JS Interop wrapper/annotation of JS exported function “Greet”

**Comment by pratikpparikh** [comment]
@jonboj I think even if dart2wasm is added for the support.  

The work that can be done here won't go to waste.  Think about from the perspective of reusability and universal platform library. Flutter require you one to develop there code in multiple native languages.  With rust you can target multiple platform with single codebase (android, ios, macos, windows, linux, and web (wasm)).  This possibility is real and highly beneficial.  

Currently if you look at the out of --web from wasm-bindgen you will notice use of global variable.  If you generate multiple of those libraries you will find that variables are overlapping and thus you either have to manually change the code and maintain that along side additional code changes in index.html.  

I hope you can see the value of why this could be great.  Yes,  Dart -> JS Interop -> wasm-bindgen -> wasm but think about it from the perspective of skipping the current generate JS and directly using Dart JS Interop code generated. It give you clean code that directly interfaces with wasm external exposed by rust.  So what is the problem, it is the following four methods that : '__wbg_new_59cb74e423758ede',__wbg_stack_558ba5917b466edd','__wbg_error_4bb6c2a97407129a','__wbindgen_object_drop_ref'

**Comment by jonboj** [comment]
@pratikpparikh 
> but think about it from the perspective of skipping the current generate JS and directly using Dart JS Interop code generated. It give you clean code that directly interfaces with wasm external exposed by rust.

Try to grab some of this scenario, but still little in a dull to me. If possible, perhaps a minimal project at github could clear up things more. Flutter brings a lot of ekstra frameworks, so maybe a minimal Dart web App. Something like a “Hello Word” Dart web App, which calls through JS Interop into a Rust wasm-bindgen snippet.

**Comment by pratikpparikh** [comment]
@jonboj I will try to put something together that helps visualize this.

**Comment by jonboj** [comment]
@pratikpparikh 
Looks like the Dart Team is enganged in work for compiling Dart to wasm https://github.com/dart-lang/sdk/issues/32894#issuecomment-831774883

**Comment by pratikpparikh** [comment]
@jonboj Thank you for sharing this information.  Maybe I should wait on them to flushing out their support before spending too much time on this.  Seems like they are going to make dart work both on client and server-side which could be exciting.
